File: packages/node/src/request.ts
import { APIConnectionError, APITimeoutError, errorFromResponse, VendorvalError } from "./errors.js";
import { generateIdempotencyKey } from "./idempotency.js";
import { decideRetryFromHeaders } from "./retry.js";
import { sleep } from "./sleep.js";
import { API_VERSION, VERSION } from "./version.js";

export type FetchLike = (input: string | URL, init?: RequestInit) => Promise<Response>;

export interface ClientOptions {
  apiKey?: string;
  baseUrl?: string;
  timeout?: number;
  maxRetries?: number;
  fetch?: FetchLike;
  /** Set false to disable client-side prefix validation (advanced). */
  validateApiKey?: boolean;
}

export interface ResolvedClientOptions {
  apiKey: string;
  baseUrl: string;
  timeout: number;
  maxRetries: number;
  fetch: FetchLike;
}

export interface RequestOptions {
  method: "GET" | "POST" | "DELETE" | "PUT" | "PATCH";
  path: string;
  query?: Record<string, string | number | boolean | undefined> | undefined;
  body?: unknown;
  /** When true, auto-inject options.idempotency_key on retry if caller didn't supply one. */
  autoIdempotency?: boolean;
  signal?: AbortSignal | undefined;
  headers?: Record<string, string> | undefined;
  /** Override the client default. */
  maxRetries?: number | undefined;
  timeout?: number | undefined;
}

export interface ApiResponse<T> {
  data: T;
  requestId: string | null;
  headers: Headers;
  status: number;
}

const DEFAULT_BASE_URL = "https://api.vendorval.com";
const DEFAULT_TIMEOUT_MS = 60_000;
const DEFAULT_MAX_RETRIES = 2;
const USER_AGENT = `vendorval-node/${VERSION} (node/${process.version.slice(1)})`;
const KEY_PREFIX = /^vv_(test|live)_/;

export function resolveOptions(opts: ClientOptions): ResolvedClientOptions {
  const apiKey = opts.apiKey ?? process.env.VENDORVAL_API_KEY ?? "";
  if (!apiKey) {
    throw new VendorvalError({
      message:
        "Missing API key. Pass `apiKey` to the Vendorval constructor or set VENDORVAL_API_KEY in the environment.",
      status: 0,
      type: "configuration_error",
      code: "missing_api_key",
      requestId: null,
    });
  }
  const validate = opts.validateApiKey !== false;
  if (validate && !KEY_PREFIX.test(apiKey)) {
    throw new VendorvalError({
      message:
        "API key has an unexpected prefix. Live keys start with `vv_live_`, test keys with `vv_test_`.",
      status: 0,
      type: "configuration_error",
      code: "invalid_api_key_prefix",
      requestId: null,
    });
  }

  const baseUrl = stripTrailingSlash(
    opts.baseUrl ?? process.env.VENDORVAL_BASE_URL ?? DEFAULT_BASE_URL,
  );

  const fetchImpl = opts.fetch ?? (globalThis.fetch as FetchLike | undefined);
  if (!fetchImpl) {
    throw new VendorvalError({
      message:
        "No fetch implementation found. Pass a `fetch` option (e.g. `node-fetch`) or use Node 18+.",
      status: 0,
      type: "configuration_error",
      code: "no_fetch",
      requestId: null,
    });
  }

  const timeout = opts.timeout ?? DEFAULT_TIMEOUT_MS;
  if (!Number.isFinite(timeout) || timeout <= 0) {
    throw new VendorvalError({
      message: "timeout must be a positive finite number of milliseconds.",
      status: 0,
      type: "configuration_error",
      code: "invalid_timeout",
      requestId: null,
    });
  }

  const maxRetries = opts.maxRetries ?? DEFAULT_MAX_RETRIES;
  if (!Number.isInteger(maxRetries) || maxRetries < 0) {
    throw new VendorvalError({
      message: "maxRetries must be a non-negative integer.",
      status: 0,
      type: "configuration_error",
      code: "invalid_max_retries",
      requestId: null,
    });
  }

  return {
    apiKey,
    baseUrl,
    timeout,
    maxRetries,
    fetch: fetchImpl,
  };
}

export async function performRequest<T>(
  client: ResolvedClientOptions,
  options: RequestOptions,
): Promise<ApiResponse<T>> {
  const url = buildUrl(client.baseUrl, options.path, options.query);
  const maxRetries = options.maxRetries ?? client.maxRetries;
  const timeout = options.timeout ?? client.timeout;

  // Generate the idempotency key (if requested) BEFORE the first attempt and
  // reuse it on every retry. If the first POST succeeded server-side but the
  // response was lost in transit, the retry sends the same key so the API
  // can deduplicate. Generating per-attempt would defeat that.
  let bodyForSend = options.autoIdempotency ? injectIdempotencyKey(options.body) : options.body;
  let lastError: VendorvalError | null = null;

  for (let attempt = 0; attempt <= maxRetries; attempt += 1) {

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout);
    const externalAbort = options.signal;
    const onExternalAbort = () => controller.abort(externalAbort?.reason);
    if (externalAbort) {
      if (externalAbort.aborted) {
        controller.abort(externalAbort.reason);
      } else {
        externalAbort.addEventListener("abort", onExternalAbort, { once: true });
      }
    }

    let response: Response;
    try {
      response = await client.fetch(url, {
        method: options.method,
        headers: buildHeaders(client.apiKey, options),
        body: serializeBody(bodyForSend, options.method),
        signal: controller.signal,
      });
    } catch (err) {
      clearTimeout(timer);
      externalAbort?.removeEventListener("abort", onExternalAbort);
      if (controller.signal.aborted && externalAbort?.aborted) {
        // Caller-driven cancellation — propagate as-is.
        throw err;
      }
      if (controller.signal.aborted) {
        lastError = new APITimeoutError(timeout);
      } else {
        lastError = new APIConnectionError(
          err instanceof Error ? err.message : "Network error",
        );
      }
      if (attempt >= maxRetries) throw lastError;
      const { delayMs } = decideRetryFromHeaders(attempt, 0, new Headers());
      await sleep(delayMs || 500 * 2 ** attempt, externalAbort);
      continue;
    } finally {
      clearTimeout(timer);
      externalAbort?.removeEventListener("abort", onExternalAbort);
    }

    const requestId = response.headers.get("x-request-id");

    if (response.ok) {
      const data = await parseBody<T>(response);
      return {
        data,
        requestId,
        headers: response.headers,
        status: response.status,
      };
    }

    const payload = await safeJson(response);
    const decision = decideRetryFromHeaders(attempt, response.status, response.headers);

    if (decision.retry && attempt < maxRetries) {
      await sleep(decision.delayMs, externalAbort);
      continue;
    }

    throw errorFromResponse({
      status: response.status,
      payload,
      headers: response.headers,
      requestId,
    });
  }

  // Unreachable: loop either returns or throws.
  throw lastError ?? new APIConnectionError("Request failed after retries");
}

function buildHeaders(apiKey: string, options: RequestOptions): HeadersInit {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${apiKey}`,
    "User-Agent": USER_AGENT,
    "X-VendorVal-API-Version": API_VERSION,
    // Opt in to the widened per-result enum
    // (`clear` / `exact_match` / `probable_match`). The API aliases
    // these down to the legacy 4-value enum for callers without the
    // header. Sending the latest version on every install dogfoods the
    // new shape; old SDK installs keep working unchanged.
    "Accept-Version": API_VERSION,
    Accept: "application/json",
    ...options.headers,
  };
  if (options.body !== undefined && options.method !== "GET") {
    headers["Content-Type"] = "application/json";
  }
  return headers;
}

function buildUrl(
  baseUrl: string,
  path: string,
  query: RequestOptions["query"],
): string {
  const url = new URL(path.startsWith("/") ? path : `/${path}`, `${baseUrl}/`);
  if (query) {
    for (const [k, v] of Object.entries(query)) {
      if (v !== undefined) url.searchParams.set(k, String(v));
    }
  }
  return url.toString();
}

function serializeBody(body: unknown, method: RequestOptions["method"]): string | undefined {
  if (method === "GET" || body === undefined) return undefined;
  return JSON.stringify(body);
}

async function parseBody<T>(response: Response): Promise<T> {
  if (response.status === 204) return undefined as T;
  const text = await response.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
}

async function safeJson(response: Response): Promise<unknown> {
  try {
    const text = await response.text();
    if (!text) return null;
    return JSON.parse(text);
  } catch {
    return null;
  }
}

function injectIdempotencyKey(body: unknown): unknown {
  if (!body || typeof body !== "object") return body;
  const cloned = { ...(body as Record<string, unknown>) };
  const existingOptions = cloned.options;
  if (existingOptions && typeof existingOptions === "object") {
    const opts = { ...(existingOptions as Record<string, unknown>) };
    if (opts.idempotency_key === undefined) {
      opts.idempotency_key = generateIdempotencyKey();
    }
    cloned.options = opts;
  } else {
    cloned.options = { idempotency_key: generateIdempotencyKey() };
  }
  return cloned;
}

function stripTrailingSlash(s: string): string {
  return s.endsWith("/") ? s.slice(0, -1) : s;
}

File: packages/python/src/vendorval_sdk/_request.py
"""Shared request plumbing for both sync and async clients."""

from __future__ import annotations

import json
import os
import platform
import re
import time
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

import httpx

from ._errors import (
    APIConnectionError,
    APITimeoutError,
    VendorvalError,
    error_from_response,
)
from ._idempotency import generate_idempotency_key
from ._retry import decide_retry
from ._version import API_VERSION, VERSION

DEFAULT_BASE_URL = "https://api.vendorval.com"
DEFAULT_TIMEOUT = 60.0
DEFAULT_MAX_RETRIES = 2

_KEY_PREFIX = re.compile(r"^vv_(test|live)_")
_USER_AGENT = (
    f"vendorval-python/{VERSION} "
    f"(python/{platform.python_version()}; {platform.python_implementation().lower()})"
)


@dataclass
class ResolvedConfig:
    api_key: str
    base_url: str
    timeout: float
    max_retries: int


@dataclass
class Prepared:
    method: str
    url: str
    headers: dict[str, str]
    json_body: Any
    timeout: float
    auto_idempotency: bool
    max_retries: int


def resolve_config(
    *,
    api_key: str | None,
    base_url: str | None,
    timeout: float | None,
    max_retries: int | None,
    validate_api_key: bool,
) -> ResolvedConfig:
    key = api_key or os.environ.get("VENDORVAL_API_KEY") or ""
    if not key:
        raise VendorvalError(
            "Missing API key. Pass api_key= or set VENDORVAL_API_KEY in the environment.",
            type="configuration_error",
            code="missing_api_key",
        )
    if validate_api_key and not _KEY_PREFIX.match(key):
        raise VendorvalError(
            "API key has an unexpected prefix. Live keys start with 'vv_live_', "
            "test keys with 'vv_test_'.",
            type="configuration_error",
            code="invalid_api_key_prefix",
        )

    url = base_url or os.environ.get("VENDORVAL_BASE_URL") or DEFAULT_BASE_URL
    url = url.rstrip("/")

    resolved_timeout = DEFAULT_TIMEOUT if timeout is None else timeout
    resolved_max_retries = DEFAULT_MAX_RETRIES if max_retries is None else max_retries
    if resolved_timeout <= 0:
        raise VendorvalError(
            "timeout must be > 0 seconds.",
            type="configuration_error",
            code="invalid_timeout",
        )
    if resolved_max_retries < 0:
        raise VendorvalError(
            "max_retries must be >= 0.",
            type="configuration_error",
            code="invalid_max_retries",
        )

    return ResolvedConfig(
        api_key=key,
        base_url=url,
        timeout=resolved_timeout,
        max_retries=resolved_max_retries,
    )


def prepare(
    cfg: ResolvedConfig,
    *,
    method: str,
    path: str,
    body: Any = None,
    query: Mapping[str, Any] | None = None,
    auto_idempotency: bool = False,
    headers: Mapping[str, str] | None = None,
    timeout: float | None = None,
    max_retries: int | None = None,
) -> Prepared:
    if not path.startswith("/"):
        path = f"/{path}"
    url = f"{cfg.base_url}{path}"
    if query:
        params = {k: str(v) for k, v in query.items() if v is not None}
        if params:
            url = str(httpx.URL(url).copy_merge_params(params))

    h = {
        "Authorization": f"Bearer {cfg.api_key}",
        "User-Agent": _USER_AGENT,
        "X-VendorVal-API-Version": API_VERSION,
        # Opt in to the widened per-result enum
        # (`clear` / `exact_match` / `probable_match`). The API aliases
        # these down to the legacy 4-value enum for callers without the
        # header. Sending the latest version on every install dogfoods
        # the new shape; old SDK installs keep working unchanged.
        "Accept-Version": API_VERSION,
        "Accept": "application/json",
    }
    if headers:
        h.update(headers)
    if body is not None and method != "GET":
        h["Content-Type"] = "application/json"

    return Prepared(
        method=method,
        url=url,
        headers=h,
        json_body=body,
        timeout=cfg.timeout if timeout is None else timeout,
        auto_idempotency=auto_idempotency,
        max_retries=cfg.max_retries if max_retries is None else max_retries,
    )


def inject_idempotency_key(body: Any) -> Any:
    if not isinstance(body, dict):
        return body
    cloned = dict(body)
    options = cloned.get("options")
    if isinstance(options, dict):
        new_opts = dict(options)
        new_opts.setdefault("idempotency_key", generate_idempotency_key())
        cloned["options"] = new_opts
    else:
        cloned["options"] = {"idempotency_key": generate_idempotency_key()}
    return cloned


@dataclass
class ApiResponse:
    data: Any
    status: int
    request_id: str | None
    headers: httpx.Headers


def _parse_body(response: httpx.Response) -> Any:
    if response.status_code == 204 or not response.content:
        return None
    try:
        return response.json()
    except json.JSONDecodeError:
        return None


def _request_id(response: httpx.Response) -> str | None:
    value = response.headers.get("x-request-id")
    return value if isinstance(value, str) else None


def execute_sync(
    client: httpx.Client,
    prepared: Prepared,
) -> ApiResponse:
    # Generate the idempotency key (if requested) before the first attempt and
    # reuse it across retries so the API can deduplicate when the first POST
    # succeeded server-side but the response was lost on the network.
    body = (
        inject_idempotency_key(prepared.json_body)
        if prepared.auto_idempotency
        else prepared.json_body
    )
    last_error: Exception | None = None
    for attempt in range(prepared.max_retries + 1):
        try:
            response = client.request(
                prepared.method,
                prepared.url,
                headers=prepared.headers,
                json=body if prepared.method != "GET" else None,
                timeout=prepared.timeout,
            )
        except httpx.TimeoutException:
            last_error = APITimeoutError(prepared.timeout)
            if attempt >= prepared.max_retries:
                raise last_error from None
            time.sleep(min(0.5 * (2 ** attempt), 30.0))
            continue
        except httpx.HTTPError as err:
            last_error = APIConnectionError(str(err))
            if attempt >= prepared.max_retries:
                raise last_error from None
            time.sleep(min(0.5 * (2 ** attempt), 30.0))
            continue

        if response.status_code < 400:
            return ApiResponse(
                data=_parse_body(response),
                status=response.status_code,
                request_id=_request_id(response),
                headers=response.headers,
            )

        delay = decide_retry(
            attempt=attempt,
            status=response.status_code,
            headers=response.headers,
        )
        if delay is not None and attempt < prepared.max_retries:
            time.sleep(delay)
            continue

        raise error_from_response(
            status=response.status_code,
            payload=_parse_body(response),
            headers=response.headers,
            request_id=_request_id(response),
        )

    if last_error is not None:
        raise last_error
    raise APIConnectionError("Request failed after retries")


async def execute_async(
    client: httpx.AsyncClient,
    prepared: Prepared,
) -> ApiResponse:
    import asyncio

    # See execute_sync — same rationale: generate the idempotency key once
    # before the loop so retries reuse it.
    body = (
        inject_idempotency_key(prepared.json_body)
        if prepared.auto_idempotency
        else prepared.json_body
    )
    last_error: Exception | None = None
    for attempt in range(prepared.max_retries + 1):
        try:
            response = await client.request(
                prepared.method,
                prepared.url,
                headers=prepared.headers,
                json=body if prepared.method != "GET" else None,
                timeout=prepared.timeout,
            )
        except httpx.TimeoutException:
            last_error = APITimeoutError(prepared.timeout)
            if attempt >= prepared.max_retries:
                raise last_error from None
            await asyncio.sleep(min(0.5 * (2 ** attempt), 30.0))
            continue
        except httpx.HTTPError as err:
            last_error = APIConnectionError(str(err))
            if attempt >= prepared.max_retries:
                raise last_error from None
            await asyncio.sleep(min(0.5 * (2 ** attempt), 30.0))
            continue

        if response.status_code < 400:
            return ApiResponse(
                data=_parse_body(response),
                status=response.status_code,
                request_id=_request_id(response),
                headers=response.headers,
            )

        delay = decide_retry(
            attempt=attempt,
            status=response.status_code,
            headers=response.headers,
        )
        if delay is not None and attempt < prepared.max_retries:
            await asyncio.sleep(delay)
            continue

        raise error_from_response(
            status=response.status_code,
            payload=_parse_body(response),
            headers=response.headers,
            request_id=_request_id(response),
        )

    if last_error is not None:
        raise last_error
    raise APIConnectionError("Request failed after retries")
